### ISSUE TITLE
bug: Detection never finished on big files

### ISSUE THREAD

**Issue by arxeiss** [aquasecurity/defsec#1108]
**Describe the bug**

When `detection.IsType` is called on big JSON (4MB etc) it never finished. It currently makes Trivy unusable for us, because it never finish. See issue https://github.com/aquasecurity/trivy/issues/3212

First, I had to clone everything and debug to find out the reason why it got stuck. 
I know it is Trivy issue, not defsec. But currently, there is no easy chance to exclude some files when we scan our codebase.

**To Reproduce**
1. Go to `pkg/detection`
2. Create test which calls `IsType("./testdata/big_file.json", bytes.NewReader(data), FileTypeAzureARM)` where big_file.json contains some string of 2M characters
3. Start

**Expected behavior**
Expect the code to finish. But it parses for ages. Because the final string is built char by char. See the code below.

**System Info**

 - tfsec version: latest master 
 - terraform version: irrelevant
 - OS: irrelevant

**Example Code**

```go
package detection

import (
	"bytes"
	"io/ioutil"
	"testing"

	"github.com/stretchr/testify/assert"
)

func Test_ArmDetection(t *testing.T) {
	t.Run("test", func(t *testing.T) {
		data, err := ioutil.ReadFile("./testdata/big_file.json")
		assert.Nil(t, err)

		res := IsType("./testdata/big_file.json", bytes.NewReader(data), FileTypeAzureARM)
		assert.False(t, res)
	})
}

```

Big file
```json
{
    "content": "...(PUT 2M characters here)..."
}
```

**Comment by arxeiss** [comment]
Bump - to remove stale label

**Comment by knqyf263** [comment]
@simar7 @giorod3 Any chance to review this PR?

**Comment by simar7** [comment]
hi @arxeiss - thanks for the report, sorry it took a while for us to get back on this. I will take a look into it shortly.

**Comment by simar7** [comment]
hi @arxeiss - I released a new version of defsec which has this change.https://github.com/aquasecurity/defsec/releases/tag/v0.84.1

Once Trivy starts to use that new version you should see some improvements in runtime.

**Comment by simar7** [comment]
> Hello @simar7 thanks for the effort, but it seems that the issue is not fixed yet.
> 
> I checkout latest branch and here are the result:
> 
> ```
> goos: linux
> goarch: amd64
> pkg: github.com/aquasecurity/defsec/pkg/detection
> cpu: Intel(R) Core(TM) i7-8665U CPU @ 1.90GHz
> BenchmarkIsType_BigFile-8   	       2	 833854810 ns/op	7122281212 B/op	  232566 allocs/op
> PASS
> ok  	github.com/aquasecurity/defsec/pkg/detection	2.559s
> ```
> 
> I was testing it with this file [big.zip](https://github.com/aquasecurity/defsec/files/10991398/big.zip) (it is JSON file inside, but GitHub does not support uploading JSON files) And it took almost 1s to parse 100kB file which is not even AzureARM

Thanks for the numbers, can you share the before and after benchmark to compare? After as in the new release (v0.84.1 or newer) vs what you see?

**Comment by arxeiss** [comment]
Hello,

I just noticed that my finding is described in another ticket. So I just copy&paste here:

Source this comment + 1 more: https://github.com/aquasecurity/trivy/issues/3212#issuecomment-1359245409

> The main issue is that the `armjson` package parses a file as JSON char by char.
> We use Trivy to check Dockerfiles inside our repo. But they are spread across multiple folders, per microservice. And it also includes `testdata` folder, where we have 4MB JSON files etc. And that takes ages when you have code like this:
> ```
> default:
> 	if c < 0x20 || c > 0x10FFFF {
> 		return nil, p.makeError("invalid unescaped character '0x%X'", c)
> 	}
> 	str += string(c)
> ```
> However, `testdata` folders are spread across whole codebase too. I know Trivy supports `--skip-dirs` but that supports only full path, not regex or something. So this mean, we cannot use latest version of Trivy, only 0.32.1 where this issue is not happening.

And the results:

**v0.84.0**:
```
goos: linux
goarch: amd64
pkg: github.com/aquasecurity/defsec/pkg/detection
cpu: Intel(R) Core(TM) i7-8665U CPU @ 1.90GHz
BenchmarkIsType_BigFile
BenchmarkIsType_BigFile-8              1        1758079612 ns/op        7122794336 B/op   232541 allocs/op
PASS
ok      github.com/aquasecurity/defsec/pkg/detection    1.764s
```

and **v0.84.1**:

```
goos: linux
goarch: amd64
pkg: github.com/aquasecurity/defsec/pkg/detection
cpu: Intel(R) Core(TM) i7-8665U CPU @ 1.90GHz
BenchmarkIsType_BigFile
BenchmarkIsType_BigFile-8              1        1894344365 ns/op        7122284720 B/op   232559 allocs/op
PASS
ok      github.com/aquasecurity/defsec/pkg/detection    1.901s
```

**Comment by simar7** [comment]
> However, testdata folders are spread across whole codebase too. I know Trivy supports --skip-dirs but that supports only full path, not regex or something. So this mean, we cannot use latest version of Trivy, only 0.32.1 where this issue is not happening.

hi @arxeiss - sorry haven't had the bandwidth to get to this but in the recent release of Trivy we added support for glob paths, which might help you in this case. 

I'll still keep this open as we can certainly improve things here but to get you unblocked you could probably try going the glob approach with `--skip-dirs` to ignore such files.

**Comment by arxeiss** [comment]
Thanks, we did some refactoring and moved all files to scan into subfolder, and running Trivy only on that subfolder.

So we are good now. But it is nice to see, that there is some progres. When our platform will grow, we might need to scan whole codebase again.

Thank you for the effort
